### PR TITLE
Merge YAMLs properly

### PR DESCRIPTION
**Pull request by tothandras**
| Q               | A
| --------------- | ---
| Bug fix?        | yes
| New feature?    | no
| API breaks?     | no
| Deprecations?   | no
| License         | Apache 2.0

### What's in this PR?

Improve the merging of values into `.pipeline.yaml`

### Why?

`mergo` is producing unexpected results when it comes to merging arbitrary JSON or YAML data.

### Checklist

- [x] Implementation tested (with at least one cloud provider)
- [x] Error handling code meets the [guideline](https://github.com/banzaicloud/pipeline/blob/master/docs/error-handling-guide.md)
- [x] Logging code meets the guideline (TODO)

### CODE REVIEW

File: spotguide/spotguide_test.go [hunk 1]
```diff
@@ -352,10 +401,10 @@ func TestCICDRepoConfigPipeline(t *testing.T) {
 	}
 
 	// Save this for debugging purposes
-	// err = ioutil.WriteFile("./actualPipeline.yaml", actualPipelineYAML, 0644)
```
Comment (orymate) [inline spotguide/spotguide_test.go]: sure?

File: spotguide/merge.go [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+	return uniqKeys
+}
+
+func merge(dst, src interface{}) (interface{}, error) {
+	if reflect.DeepEqual(dst, src) {
+		return dst, nil
+	}
+
+	switch dstV := dst.(type) {
+	// string
```
Comment (orymate) [inline spotguide/merge.go]: do we really need these comments?

File: spotguide/spotguide_test.go (hunk 1 above)
Comment (bonifaido) [reply spotguide/spotguide_test.go]: Please don't remove these lines, they are really helpful during debugging.

File: spotguide/merge.go [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+
+func maxLength(x, y int) int {
+	if x < y {
+		return y
+	}
+	return x
+}
+
+func elementOrNil(arr []interface{}, i int) interface{} {
+	if i < len(arr) {
```
Comment (sagikazarmark) [inline spotguide/merge.go]: FIY this won't work with the 0. element

File: spotguide/merge.go [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package spotguide
+
+import "reflect"
+
+func maxLength(x, y int) int {
```
Comment (sagikazarmark) [inline spotguide/merge.go]: How about math.Max?

File: spotguide/merge.go [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+func elementOrNil(arr []interface{}, i int) interface{} {
+	if i < len(arr) {
+		return arr[i]
+	}
+
+	return nil
+}
+
+func getKeys(obj map[string]interface{}) []string {
+	keys := reflect.ValueOf(obj).MapKeys()
```
Comment (sagikazarmark) [inline spotguide/merge.go]: Don't use reflect, loop through the elements instead.

File: spotguide/merge.go (hunk 2 above)
Comment (tothandras) [reply spotguide/merge.go]: I prefer to leave them, to make sure all the JSON types are handled

**Pull request merged**